File: Design Patterns/Creational/Builder/builder.cpp
#include <bits/stdc++.h>
using namespace std;

// Product Class
class Car
{
private:
  string engine;
  int wheels;
  int seats;
  string color;
  bool sunroof;
  bool navigationSystem;

  // Private constructor: Only Builder can create Car objects
  Car(const string &engine, int wheels, int seats, const string &color,
      bool sunroof, bool navigationSystem)
      : engine(engine), wheels(wheels), seats(seats),
        color(color), sunroof(sunroof), navigationSystem(navigationSystem) {}

public:
  // Friend Builder class
  class CarBuilder;

  // Getter methods
  string getEngine() const { return engine; }
  int getWheels() const { return wheels; }
  int getSeats() const { return seats; }
  string getColor() const { return color; }
  bool hasSunroof() const { return sunroof; }
  bool hasNavigationSystem() const { return navigationSystem; }

  // toString equivalent
  string toString() const
  {
    stringstream ss;
    ss << "Car [engine=" << engine
       << ", wheels=" << wheels
       << ", seats=" << seats
       << ", color=" << color
       << ", sunroof=" << (sunroof ? "true" : "false")
       << ", navigationSystem=" << (navigationSystem ? "true" : "false")
       << "]";
    return ss.str();
  }
};

// Builder Class
class Car::CarBuilder
{
private:
  string engine;
  int wheels = 4;                // Default value
  int seats = 5;                 // Default value
  string color = "Black";        // Default value
  bool sunroof = false;          // Default value
  bool navigationSystem = false; // Default value

public:
  // Builder methods (return reference for chaining)
  CarBuilder &setEngine(const string &engine)
  {
    this->engine = engine;
    return *this;
  }
  CarBuilder &setWheels(int wheels)
  {
    this->wheels = wheels;
    return *this;
  }
  CarBuilder &setSeats(int seats)
  {
    this->seats = seats;
    return *this;
  }
  CarBuilder &setColor(const string &color)
  {
    this->color = color;
    return *this;
  }
  CarBuilder &setSunroof(bool sunroof)
  {
    this->sunroof = sunroof;
    return *this;
  }
  CarBuilder &setNavigationSystem(bool navigationSystem)
  {
    this->navigationSystem = navigationSystem;
    return *this;
  }

  // Build method
  Car build()
  {
    return Car(engine, wheels, seats, color, sunroof, navigationSystem);
  }
};

// Client Code
int main()
{
  cout << "Builder Pattern Example" << endl;

  // Create a car using the builder
  Car::CarBuilder builder;
  Car car1 = builder.setEngine("V8")
                 .setColor("Red")
                 .setSeats(5)
                 .setSunroof(true)
                 .build();

  cout << car1.toString() << endl;

  // Create another car with different specifications
  Car car2 = builder.setEngine("V6")
                 .setColor("Blue")
                 .setSeats(4)
                 .build(); // Sunroof & Navigation default

  cout << car2.toString() << endl;

  return 0;
}

File: Design Patterns/Creational/Builder/builder.py
# Builder Pattern Example in Python

class Car:
    """
    Product class that represents the object we want to build.
    """

    def __init__(self, builder):
        # Assign attributes from the builder instance
        self.engine = builder.engine
        self.wheels = builder.wheels
        self.seats = builder.seats
        self.color = builder.color
        self.sunroof = builder.sunroof
        self.navigation_system = builder.navigation_system

    def __str__(self):
        """
        String representation of the Car object.
        """
        return (f"Car [engine={self.engine}, wheels={self.wheels}, seats={self.seats}, "
                f"color={self.color}, sunroof={self.sunroof}, "
                f"navigation_system={self.navigation_system}]")


class CarBuilder:
    """
    Builder class for constructing Car objects step by step.
    """

    def __init__(self):
        # Default values
        self.engine = None
        self.wheels = 4
        self.seats = 5
        self.color = "Black"
        self.sunroof = False
        self.navigation_system = False

    # Builder methods (fluent interface)
    def set_engine(self, engine):
        self.engine = engine
        return self

    def set_wheels(self, wheels):
        self.wheels = wheels
        return self

    def set_seats(self, seats):
        self.seats = seats
        return self

    def set_color(self, color):
        self.color = color
        return self

    def set_sunroof(self, sunroof):
        self.sunroof = sunroof
        return self

    def set_navigation_system(self, navigation_system):
        self.navigation_system = navigation_system
        return self

    def build(self):
        """
        Build method returns a fully constructed Car object.
        """
        return Car(self)


# Client code
if __name__ == "__main__":
    print("Builder Pattern in Python")

    # First Car
    car1 = (CarBuilder()
            .set_engine("V8")
            .set_color("Red")
            .set_seats(5)
            .set_sunroof(True)
            .build())
    print(car1)

    # Second Car
    car2 = (CarBuilder()
            .set_engine("V6")
            .set_color("Blue")
            .set_seats(4)
            .build())  # Sunroof and navigation are defaults
    print(car2)
